tools/analyze-chain.py

```diff
@@ -49,59 +49,74 @@ def run_gen(env_data: bytes, block_program_args: bytes, flags: int):
         return 117, None, 0
 
 
+def generator_references_from_indices(c, height_to_hash_cache, indices: List[int]) -> List[bytes]:
+    ref_block_blobs = []
+    for h in indices:
+        if h not in height_to_hash_cache:
+            ref = c.execute("SELECT header_hash FROM full_blocks WHERE height=?", (h,))
+            height_to_hash_cache[h] = ref.fetchone()[0]
+            ref.close()
+
+        ref = c.execute("SELECT block FROM full_blocks WHERE header_hash=?", (height_to_hash_cache[h],))
+        ref_block = FullBlock.from_bytes(zstd.decompress(ref.fetchone()[0]))
+        ref_block_blobs.append(bytes(ref_block.transactions_generator))
+        ref.close()
+    return ref_block_blobs
+
+
 @click.command()
 @click.argument("file", type=click.Path(), required=True)
 @click.option(
     "--mempool-mode", default=False, is_flag=True, help="execute all block generators in the strict mempool mode"
 )
-def main(file: Path, mempool_mode: bool):
+@click.option("--precache", default=False, is_flag=True, help="pre-cache block heights")
+@click.option("--start", default=225000, help="first block to examine")
+@click.option("--end", default=None, help="last block to examine")
+def main(file: Path, mempool_mode: bool, start: int, end: Optional[int], precache: bool):
     c = sqlite3.connect(file)
 
-    rows = c.execute("SELECT header_hash, height, block FROM full_blocks ORDER BY height")
+    height_to_hash_cache = {}
+
+    rows = c.execute("SELECT max(height) FROM full_blocks")
+    (max_height,) = next(rows)
+    end_height = max_height if end is None else min(max_height, end)
 
-    height_to_hash: List[bytes] = []
+    if precache:
+        rows = c.execute(f"SELECT header_hash, height FROM full_blocks WHERE height <= {end_height}")
+        for hh, height in rows:
+            height_to_hash_cache[height] = hh
+            print(f"precaching {height}         \r", end="")
```

Comment:
I think we should remove the `height_to_hash_cache` entirely. I don't think we need it anymore, if we can just require a v2 blockchain database instead